File: tpcp-go/tpcp/dlq.go
package tpcp

// DLQ is a Dead Letter Queue backed by a buffered channel (capacity 100).
// Envelopes that have no registered handler are deposited here.
type DLQ struct {
	ch chan *TPCPEnvelope
}

// NewDLQ creates a DLQ with capacity 100.
func NewDLQ() *DLQ {
	return &DLQ{ch: make(chan *TPCPEnvelope, 100)}
}

// Enqueue adds an envelope to the queue. Returns false if the queue is full.
func (q *DLQ) Enqueue(env *TPCPEnvelope) bool {
	select {
	case q.ch <- env:
		return true
	default:
		return false
	}
}

// Drain removes and returns all envelopes currently in the queue.
func (q *DLQ) Drain() []*TPCPEnvelope {
	var out []*TPCPEnvelope
	for {
		select {
		case env := <-q.ch:
			out = append(out, env)
		default:
			return out
		}
	}
}

// Len returns the number of envelopes currently queued.
func (q *DLQ) Len() int {
	return len(q.ch)
}

File: tpcp-java/src/main/java/io/tpcp/core/DLQ.java
package io.tpcp.core;

import io.tpcp.schema.TPCPEnvelope;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.LinkedBlockingQueue;

/**
 * Dead Letter Queue for unhandled TPCP envelopes.
 * Capacity capped at 100 to prevent unbounded growth.
 */
public class DLQ {
    private final LinkedBlockingQueue<TPCPEnvelope> queue = new LinkedBlockingQueue<>(100);

    /** Adds an envelope to the queue. Returns false if the queue is full. */
    public boolean enqueue(TPCPEnvelope envelope) {
        return queue.offer(envelope);
    }

    /** Removes and returns all envelopes currently in the queue. */
    public List<TPCPEnvelope> drain() {
        List<TPCPEnvelope> result = new ArrayList<>();
        queue.drainTo(result);
        return result;
    }

    /** Returns the number of envelopes in the queue. */
    public int size() {
        return queue.size();
    }
}
